govcd/nsxt_edgegateway_test.go

```diff
@@ -197,4 +215,224 @@ func (vcd *TestVCD) Test_NsxtEdgeVdcGroup(check *C) {
+
+	// Edge Gateway always allocates 1 IP as its primary
+	check.Assert(usedIPs[0].IPAddress, Equals, "1.1.1.3")
+	check.Assert(usedIPs[0].Category, Equals, "PRIMARY_IP")
+
+	// Attempt to get 1 unallocated IP
+	ipAddr, err := createdEdge.GetUnusedExternalIPAddresses(1, netip.Prefix{}, false)
+	check.Assert(err, IsNil)
+	ipsCompared := compareEachIpElementAndOrder(ipAddr, []netip.Addr{netip.MustParseAddr("1.1.1.4")})
+	check.Assert(ipsCompared, Equals, true)
+
+	// Attempt to get 10 unallocated IPs
+	ipAddr, err = createdEdge.GetUnusedExternalIPAddresses(10, netip.Prefix{}, true)
+	check.Assert(err, IsNil)
+	ipsCompared = compareEachIpElementAndOrder(ipAddr, []netip.Addr{
+		netip.MustParseAddr("1.1.1.4"),
+		netip.MustParseAddr("1.1.1.5"),
+		netip.MustParseAddr("1.1.1.6"),
+		netip.MustParseAddr("1.1.1.7"),
+		netip.MustParseAddr("1.1.1.8"),
+		netip.MustParseAddr("1.1.1.9"),
+		netip.MustParseAddr("1.1.1.10"),
+		netip.MustParseAddr("1.1.1.11"),
+		netip.MustParseAddr("1.1.1.12"),
+		netip.MustParseAddr("1.1.1.13"),
+	})
+	check.Assert(ipsCompared, Equals, true)
+
+	// Attempt to get IP but filter it off by prefix
+	ipAddr, err = createdEdge.GetUnusedExternalIPAddresses(1, netip.MustParsePrefix("192.168.1.1/24"), false)
+	// Expect an error because Edge Gateway does not have IPs from required subnet 192.168.1.1/24
+	check.Assert(err, NotNil)
+	check.Assert(ipAddr, IsNil)
+
+	// Attempt to get all unused IPs
+	allIps, err := createdEdge.GetAllUnusedExternalIPAddresses(true)
+	check.Assert(err, IsNil)
+	ipsCompared = compareEachIpElementAndOrder(allIps, []netip.Addr{
+		netip.MustParseAddr("1.1.1.4"),
+		netip.MustParseAddr("1.1.1.5"),
+		netip.MustParseAddr("1.1.1.6"),
+		netip.MustParseAddr("1.1.1.7"),
+		netip.MustParseAddr("1.1.1.8"),
+		netip.MustParseAddr("1.1.1.9"),
+		netip.MustParseAddr("1.1.1.10"),
+		netip.MustParseAddr("1.1.1.11"),
+		netip.MustParseAddr("1.1.1.12"),
+		netip.MustParseAddr("1.1.1.13"),
+		netip.MustParseAddr("1.1.1.14"),
+		netip.MustParseAddr("1.1.1.15"),
+		netip.MustParseAddr("1.1.1.16"),
+		netip.MustParseAddr("1.1.1.17"),
+		netip.MustParseAddr("1.1.1.18"),
+		netip.MustParseAddr("1.1.1.19"),
+		netip.MustParseAddr("1.1.1.20"),
+		netip.MustParseAddr("1.1.1.21"),
+		netip.MustParseAddr("1.1.1.22"),
+		netip.MustParseAddr("1.1.1.23"),
+		netip.MustParseAddr("1.1.1.24"),
+		netip.MustParseAddr("1.1.1.25"),
+	})
+	check.Assert(ipsCompared, Equals, true)
+	check.Assert(len(allIps), Equals, 22)
+
+	// Verify that GetAllocatedIpCount returns correct number of allocated IPs
+	totalAllocationIpCount, err := createdEdge.GetAllocatedIpCount(true)
+	check.Assert(err, IsNil)
+	check.Assert(totalAllocationIpCount, NotNil)
+	check.Assert(totalAllocationIpCount, Equals, 23) // 22 unused IPs + 1 primary
+
+	// Try to deallocate more IPs than allocated
+	failedDeallocationIpCount, err := createdEdge.QuickDeallocateIpCount(24)
+	check.Assert(err, NotNil)
+	check.Assert(failedDeallocationIpCount, IsNil)
+
```

Comment:
At this point, we should check that the previous operation has failed without side effects, i.e. that the expected IPs are still there. (If I understand correctly how it should work)

Reply:
I have added that test